src/main/kotlin/io/emeraldpay/dshackle/rpc/SubscribeNodeStatus.kt

```diff
@@ -0,0 +1,125 @@
+    private val multistreams: CurrentMultistreamHolder
+) {
+
+    fun subscribe(req: Mono<SubscribeNodeStatusRequest>): Flux<NodeStatusResponse> =
+        req.flatMapMany {
+            val knownUpstreams = ConcurrentHashMap<String, Boolean>()
+            // send known upstreams details immediately
+            val descriptions = Flux.fromIterable(
+                multistreams.all()
+                    .flatMap { multiStream ->
+                        multiStream.getAll().map { up ->
+                            NodeStatusResponse.newBuilder()
+                                .setNodeId(up.getId())
+                                .setDescription(buildDescription(multiStream.chain, up))
+                                .setStatus(buildStatus(up.getStatus(), up.getHead().getCurrentHeight()))
+                                .build()
+                        }
+                    }
+            )
+
+            // subscribe on head/status updates for known upstreams
+            val upstreamUpdates = Flux.merge(
+                multistreams.all()
+                    .flatMap { ms ->
+                        ms.getAll().map { up ->
+                            knownUpstreams[up.getId()] = true
+                            subscribeUpstreamUpdates(ms.chain, up)
+                        }
+                    }
+            )
+
+            // subscribe on head/status updates for just added upstreams
+            val muliStreamUpdates = Flux.merge(
+                multistreams.all()
+                    .map { ms ->
+                        ms.subscribeAddedUpstreams()
+                            .distinctUntilChanged {
+                                it.getId()
+                            }
+                            .filter { knownUpstreams[it.getId()] != true }.flatMap {
+                                knownUpstreams[it.getId()] = true
+                                Flux.concat(
+                                    Mono.just(
+                                        NodeStatusResponse.newBuilder()
+                                            .setNodeId(it.getId())
+                                            .setDescription(buildDescription(ms.chain, it))
+                                            .setStatus(buildStatus(it.getStatus(), it.getHead().getCurrentHeight()))
+                                            .build()
+                                    ),
+                                    subscribeUpstreamUpdates(ms.chain, it)
+                                )
+                            }
+                    }
+            )
+
+            Flux.concat(descriptions, Flux.merge(upstreamUpdates, muliStreamUpdates))
+        }
+
+    private fun subscribeUpstreamUpdates(chain: Chain, upstream: Upstream): Flux<NodeStatusResponse> {
+        val r = Sinks.many().multicast().directBestEffort<Boolean>()
+        val heads = Mono.just(upstream).repeatWhen { r.asFlux() }.sample(Duration.ofSeconds(10)).flatMap { up ->
```

Comment:
let's parametrize sample by request may be?

Reply:
it's a period of time to re-subscribe to the head

Reply:
sample allows to skip extra events in a time duration and always produces the last one - that's what we need)